### DISCUSSION TITLE
Can I have Basic Authentication with 2FA or OAuth2.0 flow?

### DISCUSSION THREAD

**Discussion by emregency** [directus/directus#3679] [Archive]
When I enable 2FA and try to login through Azure AD, this is what I get:

`{"errors":[{"message":"\"otp\" is required","extensions":{"code":"INVALID_OTP"}}]}`

Is it possible to have both the option of `Basic Authentication + 2FA` or an `OAuth` flow without the 2FA extension?

**Comment by rijkvanzanten** [comment]
> Is it possible to have both the option of Basic Authentication + 2FA or an OAuth flow without the 2FA extension?

Not as of right now

**Reply by emregency** [reply]
Thanks for the quick answer.

**Reply by rijkvanzanten** [reply]
@emregency Do you think logging in through SSO should allow you to "pass" 2FA? 🤔

**Reply by emregency** [reply]
So currently I cannot have 2FA with SSO neither as I am redirected to my providers page without having the option to pass an OTP to my oAurh provider. 

Generally speaking I would see every authentication method as independent flows. If I am not wrong Auth0 implements them orthogonal: it’s either 2FA, either magic link, either SSO. 

In our case, if an entire project (tenant) has an enhanced security policy with mandatory 2FA, then maybe at the end of each flow we may have an extra OTP step. 

That’s how I would imagine it.

**Reply by rijkvanzanten** [reply]
> In our case, if an entire project (tenant) has an enhanced security policy with mandatory 2FA, then maybe at the end of each flow we may have an extra OTP step.

This is what I was wondering about. Should we show Directus' 2FA input after signing in with SSO as well, or should just SSO be enough? 🤔

**Reply by emregency** [reply]
🤔 Currently, in Directus 2FA is enabled per user basis. So there is no centralized enforcement option. If so, IMO the authentication options should not be concatenated: “SSO w/o extra Directus 2FA” or “password optional. 2FA” scenarii make more sense as the current implementation is per user.

i.e. 2FA app for mail+password is 😎 but I can have a phone call, sms / mail OTP, Yubikey 2FA in my oAuth flow.

**Reply by benhaynes** [reply]
Here are my thoughts:

* Let's rename `2FA` to `MFA` (Multi-Factor Auth). This has the added (future) benefit of allowing users/admins to choose how many factors are required: 1, 2, 3... etc
* I think MFA should be extensible, so that you can choose from Authenticators, Email Codes, SMS Codes, etc (new extension type)
* While I think SSO should be considered "1 Factor", I think it's ideal to just make this configurable. Then we can have a security toggle that says "Disable MFA for SSO"... or something like that.

That should give complete flexibility for whatever the project needs. Now we just need to find time to actually implement this. In the meantime, we'll push a patch to stop the issue mentioned here.

**Reply by emregency** [reply]
I like the extend of the scope @benhaynes 👍. It gives all the options but it’s quite labor intensive as you mentioned.

I really like the “disable MFA for SSO” concept as right now we cannot have SSO and 2FA at the same time.

**Reply by emregency** [reply]
Hey guys,

What do you think about a simple "disable MFA for SSO"? Do you think it's worth a try?

**Reply by rijkvanzanten** [reply]
Will that be a project-setting, role-setting or per-user?

**Reply by emregency** [reply]
It sounds more like a principle so I would see it as a project setting. Does that makes sense?

**Reply by rijkvanzanten** [reply]
Yeah! Also wondering the other way: do you ever _want_ Directus' MFA when you login using a third-party service? 🤔

**Reply by emregency** [reply]
I would say no because it’s either or. Either I go with Directus login flow or with a third-party login flow. Would you agree to that?

**Reply by rijkvanzanten** [reply]
I was thinking the same thing!
